### PR TITLE
Revamp the documentation of the root of core

### PR DESCRIPTION
**Pull request by tomaka**
They are totally obsolete, so let's make them less complete and less obsolete.

### CODE REVIEW

**Comment by tomaka** [comment]
Thinking about it, I think the documentation of `libp2p` should explain how to use it from an end-user perspective (eg. if you want to use floodsub/kademlia/...), while the documentation of `libp2p-core` should explain how to create a custom protocol or a muxer.

File: core/src/lib.rs [hunk 1]
```diff
@@ -23,145 +23,41 @@
...
-//! also implements the `ConnectionUpgrade` trait and will choose one of the protocols amongst the
-//! ones supported.
+//! # General overview
+//!
+//! From a user point of view, the main concepts of libp2p are:
+//!
+//! - A `PeerId` is a unique global identifier for a node on the network. Each node should have a
+//!   different `PeerId`. Normally, a `PeerId` is the hash of the public key used to negotiate
+//!   encryption on the communication channel, thereby guaranteeing that they cannot be spoofed.
+//! - The `Transport` trait define how to reach a remote or listen for incoming remote connections.
```
Comment (dvdplm) [inline core/src/lib.rs]:
```suggestion
//! - The `Transport` trait defines how to reach a remote node or listen for incoming remote connections.
```

File: core/src/lib.rs [hunk 2]
```diff
@@ -23,145 +23,41 @@
...
-//! ## Grouping protocols
-//!
-//! You can use the `.or_upgrade()` method to group multiple upgrades together. The return value
-//! also implements the `ConnectionUpgrade` trait and will choose one of the protocols amongst the
-//! ones supported.
+//! # General overview
+//!
+//! From a user point of view, the main concepts of libp2p are:
+//!
+//! - A `PeerId` is a unique global identifier for a node on the network. Each node should have a
```
Comment (dvdplm) [inline core/src/lib.rs]:
```suggestion
//! - A `PeerId` is a unique global identifier for a node on the network. Each node must have a
```

(I think it's mandatory for peer ids to be unique?)

File: core/src/lib.rs [hunk 3]
```diff
@@ -23,145 +23,41 @@
...
-//! let _ = rt.block_on(ping_dialer).unwrap();
-//! # }
-//! ```
-//!
-//! ## Grouping protocols
-//!
-//! You can use the `.or_upgrade()` method to group multiple upgrades together. The return value
-//! also implements the `ConnectionUpgrade` trait and will choose one of the protocols amongst the
-//! ones supported.
+//! # General overview
```
Comment (dvdplm) [inline core/src/lib.rs]:
```suggestion
//! # Overview
```

File: core/src/lib.rs [hunk 4]
```diff
@@ -23,145 +23,41 @@
...
+//!
+//! From a user point of view, the main concepts of libp2p are:
+//!
+//! - A `PeerId` is a unique global identifier for a node on the network. Each node should have a
+//!   different `PeerId`. Normally, a `PeerId` is the hash of the public key used to negotiate
+//!   encryption on the communication channel, thereby guaranteeing that they cannot be spoofed.
+//! - The `Transport` trait define how to reach a remote or listen for incoming remote connections.
+//!   See the `transport` module.
+//! - The `Swarm` struct contains all the active or pending connections to remotes, including all
+//!   the opened protocols.
```
Comment (dvdplm) [inline core/src/lib.rs]: Not sure I understand what you mean by "including all the opened protocols". Worth clarifying/extending imo.

File: core/src/lib.rs [hunk 5]
```diff
@@ -23,145 +23,41 @@
...
+//! - The `NetworkBehaviour` trait is implemented on types that can customize the behaviour of the
```
Comment (dvdplm) [inline core/src/lib.rs]: Maybe:
```suggestion
//! - Use the `NetworkBehaviour` trait to customize the behaviour of a `Swarm`. It is the `NetworkBehaviour` that controls what happens on the network. Multiple types that implement `NetworkBehaviour` can be composed into a single behaviour.
```

File: core/src/lib.rs [hunk 6]
```diff
@@ -23,145 +23,41 @@
...
+//!   `Swarm`. It is the `NetworkBehaviour` that defines what happens on the network.
+//! - The `Topology` trait is implemented on types that can hold the layout of the network.
```
Comment (dvdplm) [inline core/src/lib.rs]: Maybe:
```suggestion
//! - The `Topology` trait is implemented for types that need to know about the layout of the network to operate.
```

(I guess my underlying question is: those types that need to know about the surrounding topology actually **must** use the `Topology` trait right? For those cases it's not really optional, which is what `can` sort of implies.)

File: core/src/lib.rs [hunk 7]
```diff
@@ -23,145 +23,41 @@
...
+//!
+//! In order to get started with libp2p, proceed with the following steps:
+//!
+//! - Create an instance of a type that implements `Transport` based on the list of transport
+//!   protocols (such as TCP/IP, DNS, bluetooth, etc.) that you want to support. See the
```
Comment (dvdplm) [inline core/src/lib.rs]:
```suggestion
//!   protocols that you want to support (such as TCP/IP, DNS, bluetooth, etc.). See the
```

File: core/src/lib.rs [hunk 8]
```diff
@@ -23,145 +23,41 @@
...
+//! - Create an instance of a type that implements `NetworkBehaviour`. See the `swarm` module for
+//!   information about how to do that.
+//! - Create an instance of a swarm that implements `Topology`. See the `topology` module for
+//!   information about how to do that.
+//! - Create a `Swarm` by passing the three objects you have just created.
+//!
+//! Additionally, if you want to create your own protocol, there are some additional concepts:
+//!
+//! - The `StreamMuxer` trait is implemented on structs that hold a connection to a remote and can
+//!   subdivide this connection into multiple substreams. See the `muxing` module.
```
Comment (dvdplm) [inline core/src/lib.rs]: I'd mention where `StreamMuxer` fits in among the three objects mentioned above, something like "Muxers are used with handlers, which are used with `Swarm`" (but put better)?

File: core/src/lib.rs [hunk 9]
```diff
@@ -23,145 +23,41 @@
...
+//! - Create an instance of a type that implements `Transport` based on the list of transport
+//!   protocols (such as TCP/IP, DNS, bluetooth, etc.) that you want to support. See the
+//!   `transport` module for information about how to do that.
+//! - Create an instance of a type that implements `NetworkBehaviour`. See the `swarm` module for
+//!   information about how to do that.
+//! - Create an instance of a swarm that implements `Topology`. See the `topology` module for
+//!   information about how to do that.
+//! - Create a `Swarm` by passing the three objects you have just created.
+//!
+//! Additionally, if you want to create your own protocol, there are some additional concepts:
```
Comment (dvdplm) [inline core/src/lib.rs]:
```suggestion
//! If you want to create your own protocol, there are some additional lower-level concepts:
```

File: core/src/lib.rs [hunk 10]
```diff
@@ -23,145 +23,41 @@
...
+//!
+//! - The `StreamMuxer` trait is implemented on structs that hold a connection to a remote and can
+//!   subdivide this connection into multiple substreams. See the `muxing` module.
+//! - The `UpgradeInfo`, `InboundUpgrade` and `OutboundUpgrade` traits define how to upgrade each
+//!   individual substream to use a protocol. See the `upgrade` module.
+//! - The `ProtocolsHandler` trait defines how each active connection to a remote should behave:
+//!   how to handle incoming substreams, which protocols are supported, when to open a new
+//!   outbound substream, etc. See the `protocols_handler` trait.
```
Comment (dvdplm) [inline core/src/lib.rs]:
```suggestion
//!   outbound substream, etc. Used inside the `NetworkBehaviour`. See the `protocols_handler` trait.
```

File: core/src/lib.rs [hunk 11]
```diff
@@ -23,145 +23,41 @@
...
-//! # }
-//! ```
-//!
-//! ## Grouping protocols
-//!
-//! You can use the `.or_upgrade()` method to group multiple upgrades together. The return value
-//! also implements the `ConnectionUpgrade` trait and will choose one of the protocols amongst the
-//! ones supported.
+//! # General overview
+//!
```
Comment (dvdplm) [inline core/src/lib.rs]: Not sure where it fits best, but I think it'd be good to mention the "compos-ability" philosophy of rust-libp2p, where the various components are built to be as simple as they can be and then combined into more complex shapes. I think it's a cool feature of the library (so worth mentioning in its own right) but it also something that is good to know when reading the docs and figuring out how it all fits together.

File: core/src/lib.rs (hunk 2 above)
Comment (tomaka) [reply core/src/lib.rs]: It's more like "each node should have a different one, if they don't then ouch"

File: core/src/lib.rs (hunk 4 above)
Comment (tomaka) [reply core/src/lib.rs]: As in "manages the state of all the substreams that have been opened, and all the upgrades that were built upon these substreams"

File: core/src/lib.rs (hunk 6 above)
Comment (tomaka) [reply core/src/lib.rs]: The types that implement `Topology` are just sort of glorified hash maps. You can say that they *hold* the layout of the network. When something needs to know the layout, it gets passed a reference to an implementation of `Topology`.

File: core/src/lib.rs (hunk 4 above)
Comment (dvdplm) [reply core/src/lib.rs]: That is much better, so "The `Swarm` struct contains all active and pending connections to remotes and manages the state of all the substreams that have been opened, and all the upgrades that were built upon these substreams".

File: core/src/lib.rs (hunk 6 above)
Comment (dvdplm) [reply core/src/lib.rs]: Ok, so how about this: "- The `Topology` trait is implemented for types that hold the layout of a network. When other components need the network layout to operate, they are passed an instance of `Topology`."?

**Comment by tomaka** [comment]
I merged the paragraphs because this is the documentation of `libp2p-core`, therefore for people who are interested in the lower-level concepts.

**Pull request merged**